spring-ai-pinecone/src/main/java/org/springframework/ai/vectorstores/pinecone/vectorstores/PineconeVectorStore.java

```diff
@@ -0,0 +1,196 @@
+							com.google.protobuf.Value.newBuilder()
+								.setStringValue(documents.get(i).getMetadata().toString())
+								.build())
+					.build())
+				// The ID
+				.setId(documents.get(i).getId())
+				.build());
+		}
+		UpsertRequest request = UpsertRequest.newBuilder()
+			.addAllVectors(upsertVectors)
+			// TODO: set namespace
+			.setNamespace(pineconeNamespace)
+			.build();
+
+		UpsertResponse upsertResponse = pineconeConnection.getBlockingStub().upsert(request);
+		logger.info("Upserted {} records", upsertResponse.getUpsertedCount());
+
+	}
+
+	// Delete documents from the vector store using the injected PineconeClient
+	@Override
+	public Optional<Boolean> delete(List<String> idList) {
+		DeleteRequest deleteRequest = DeleteRequest.newBuilder()
+			.setNamespace(pineconeNamespace)
+			.addAllIds(idList)
+			.setDeleteAll(false)
+			.build();
+
+		DeleteResponse deleteResponse = pineconeConnection.getBlockingStub().delete(deleteRequest);
+		logger.info("Deleted {} records", deleteResponse.getSerializedSize());
+
+		// TODO: check if delete was successful, is the deleteResponse.getSerializedSize()
+		// > 0 a check we can use to confirm this ?
+		return Optional.of(true);
+	}
+
+	@Override
+	public List<Document> similaritySearch(String query) {
+		return similaritySearch(query, 10); // TODO: Discuss this - default k = 10 if not
+											// specified
+	}
+
+	@Override
+	public List<Document> similaritySearch(String query, int k) {
+		// use the Embeddings Client to get the query embedding
+		List<Double> userQueryEmbedding = this.embeddingClient.embed(query);
+		// convert the query embedding from List(Double) to List(Float)
+		List<Float> userQueryEmbeddingFloat = new ArrayList<>();
+
+		// use the Pinecone Lib to get the similarity search results
+		QueryRequest queryRequest = QueryRequest.newBuilder()
+			.addAllVector(userQueryEmbeddingFloat)
+			.setNamespace(pineconeNamespace)
+			.setTopK(2)
+			.setIncludeMetadata(true)
+			.setTopK(k)
+			.build();
+
+		QueryResponse queryResponse = pineconeConnection.getBlockingStub().query(queryRequest);
+		List<SingleQueryResults> resList = queryResponse.getResultsList();
+		List<Document> docList = new ArrayList<>();
+		for (SingleQueryResults res : resList) {
+			res.getMatchesList().forEach(match -> {
+				String id = match.getId();
+				Float score = match.getScore();
+
+				Map<String, Object> metadata = new HashMap<>();
+				// iterate through match.getMetadata().getFieldsMap() and create a
+				// Map<String, Object> metadata
+				for (Map.Entry<String, com.google.protobuf.Value> entry : match.getMetadata()
+					.getFieldsMap()
+					.entrySet()) {
+					String key = entry.getKey();
+					com.google.protobuf.Value value = entry.getValue();
+					metadata.put(key, value);
+				}
+				// TODO: Score is not returned, we need to find a way to get it
+				Document doc = new Document(id, metadata);
+				docList.add(doc);
+			});
+		}
+		return docList;
+	}
```

Comment:
This looks like mostly duplicate code to `public List<Document> similaritySearch(String query, int k, double threshold)`. Can this be refactored to avoid duplication?